# Importing required libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt
import math

# Importing the data

In [2]:
df = pd.read_csv('storedata_total.csv')
df.head()

custid  retained     created  firstorder   lastorder  esent   eopenrate  \
0  6H6T6N         0  28-09-2012  11-08-2013  11-08-2013     29  100.000000   
1  APCENR         1  19-12-2010  01-04-2011  19-01-2014     95   92.631579   
2  7UP6MS         0  03-10-2010  01-12-2010  06-07-2011      0    0.000000   
3  7ZEW8G         0  22-10-2010  28-03-2011  28-03-2011      0    0.000000   
4  8V726M         1  27-11-2010  29-11-2010  28-01-2013     30   90.000000   

   eclickrate  avgorder   ordfreq  paperless  refill  doorstep     favday city  
0    3.448276     14.52  0.000000          0       0         0     Monday  DEL  
1   10.526316     83.69  0.181641          1       1         1     Friday  DEL  
2    0.000000     33.58  0.059908          0       0         0  Wednesday  DEL  
3    0.000000     54.96  0.000000          0       0         0   Thursday  BOM  
4   13.333333    111.91  0.008850          0       0         0     Monday  BOM

In [3]:
df.shape

(30801, 15)

In [4]:
df.dtypes

custid         object
retained        int64
created        object
firstorder     object
lastorder      object
esent           int64
eopenrate     float64
eclickrate    float64
avgorder      float64
ordfreq       float64
paperless       int64
refill          int64
doorstep        int64
favday         object
city           object
dtype: object

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30801 entries, 0 to 30800
Data columns (total 15 columns):
custid        30781 non-null object
retained      30801 non-null int64
created       30781 non-null object
firstorder    30781 non-null object
lastorder     30781 non-null object
esent         30801 non-null int64
eopenrate     30801 non-null float64
eclickrate    30801 non-null float64
avgorder      30801 non-null float64
ordfreq       30801 non-null float64
paperless     30801 non-null int64
refill        30801 non-null int64
doorstep      30801 non-null int64
favday        30801 non-null object
city          30801 non-null object
dtypes: float64(4), int64(5), object(6)
memory usage: 3.5+ MB


In [6]:
df.describe(include=np.number)

retained         esent     eopenrate    eclickrate      avgorder  \
count  30801.000000  30801.000000  30801.000000  30801.000000  30801.000000   
mean       0.794520     28.138405     25.554838      5.671139     61.873944   
std        0.404059     16.751380     29.557106     10.561704     41.007588   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        1.000000     16.000000      2.040816      0.000000     40.020000   
50%        1.000000     32.000000     13.207547      0.000000     50.970000   
75%        1.000000     42.000000     40.000000      7.142857     74.280000   
max        1.000000    291.000000    100.000000    100.000000   2600.140000   

            ordfreq     paperless        refill      doorstep  
count  30801.000000  30801.000000  30801.000000  30801.000000  
mean       0.037729      0.649037      0.095094      0.038895  
std        0.103932      0.477279      0.293350      0.193347  
min        0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000      0.000000  
50%        0.000000      1.000000      0.000000      0.000000  
75%        0.040816      1.000000      0.000000      0.000000  
max        3.250000      1.000000      1.000000      1.000000

In [7]:
df.describe(include='O')

custid     created  firstorder   lastorder  favday   city
count    30781       30781       30781       30781   30801  30801
unique   30769        2821        2672        2414       7      4
top     GAWC5Y  28-08-2012  17-09-2013  13-01-2014  Monday    BOM
freq         2         928         197         339    6762  11566

In [8]:
df.describe(include='all')

custid      retained     created  firstorder   lastorder  \
count    30781  30801.000000       30781       30781       30781   
unique   30769           NaN        2821        2672        2414   
top     GAWC5Y           NaN  28-08-2012  17-09-2013  13-01-2014   
freq         2           NaN         928         197         339   
mean       NaN      0.794520         NaN         NaN         NaN   
std        NaN      0.404059         NaN         NaN         NaN   
min        NaN      0.000000         NaN         NaN         NaN   
25%        NaN      1.000000         NaN         NaN         NaN   
50%        NaN      1.000000         NaN         NaN         NaN   
75%        NaN      1.000000         NaN         NaN         NaN   
max        NaN      1.000000         NaN         NaN         NaN   

               esent     eopenrate    eclickrate      avgorder       ordfreq  \
count   30801.000000  30801.000000  30801.000000  30801.000000  30801.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean       28.138405     25.554838      5.671139     61.873944      0.037729   
std        16.751380     29.557106     10.561704     41.007588      0.103932   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%        16.000000      2.040816      0.000000     40.020000      0.000000   
50%        32.000000     13.207547      0.000000     50.970000      0.000000   
75%        42.000000     40.000000      7.142857     74.280000      0.040816   
max       291.000000    100.000000    100.000000   2600.140000      3.250000   

           paperless        refill      doorstep  favday   city  
count   30801.000000  30801.000000  30801.000000   30801  30801  
unique           NaN           NaN           NaN       7      4  
top              NaN           NaN           NaN  Monday    BOM  
freq             NaN           NaN           NaN    6762  11566  
mean        0.649037      0.095094      0.038895     NaN    NaN  
std         0.477279      0.293350      0.193347     NaN    NaN  
min         0.000000      0.000000      0.000000     NaN    NaN  
25%         0.000000      0.000000      0.000000     NaN    NaN  
50%         1.000000      0.000000      0.000000     NaN    NaN  
75%         1.000000      0.000000      0.000000     NaN    NaN  
max         1.000000      1.000000      1.000000     NaN    NaN

# Missing Value Imputation

In [9]:
df.isnull().sum()

custid        20
retained       0
created       20
firstorder    20
lastorder     20
esent          0
eopenrate      0
eclickrate     0
avgorder       0
ordfreq        0
paperless      0
refill         0
doorstep       0
favday         0
city           0
dtype: int64

In [10]:
round(df.isnull().sum()/len(df)*100 ,2)

custid        0.06
retained      0.00
created       0.06
firstorder    0.06
lastorder     0.06
esent         0.00
eopenrate     0.00
eclickrate    0.00
avgorder      0.00
ordfreq       0.00
paperless     0.00
refill        0.00
doorstep      0.00
favday        0.00
city          0.00
dtype: float64

In [11]:
df1 = df.copy()
df1.dropna(inplace=True)
df1.reset_index(drop=True,inplace=True)

In [12]:
df1.shape

(30781, 15)

# Handling non-standard data and Data type Conversions

In [13]:
df1.dtypes

custid         object
retained        int64
created        object
firstorder     object
lastorder      object
esent           int64
eopenrate     float64
eclickrate    float64
avgorder      float64
ordfreq       float64
paperless       int64
refill          int64
doorstep        int64
favday         object
city           object
dtype: object

#### From the above cell, it is seen that the columns created, firstorder and lastorder are supposed to be date columns. Also, there are some date values which are out of range like 1/0/00 in the columns firstorder and lastorder. All these non-standard data are dropped

In [14]:
non_ms_values = df1[(df1['firstorder']=='1/0/00') | (df1['lastorder']=='1/0/00')].index
df2 = df1.drop(non_ms_values)
df2.reset_index(drop=True,inplace=True)
# shape of our dataset after dropping the non-standard data
df2.shape

(30758, 15)

#### Converting datatype of the columns 'created', 'firstorder' and 'lastorder' to datetime format

In [15]:
# converting the columns to datetime format
df2['created'] = df2['created'].apply(lambda x: pd.to_datetime(str(x)))
df2['firstorder'] = df2['firstorder'].apply(lambda x: pd.to_datetime(str(x)))
df2['lastorder'] = df2['lastorder'].apply(lambda x: pd.to_datetime(str(x)))

In [16]:
df2.dtypes

custid                object
retained               int64
created       datetime64[ns]
firstorder    datetime64[ns]
lastorder     datetime64[ns]
esent                  int64
eopenrate            float64
eclickrate           float64
avgorder             float64
ordfreq              float64
paperless              int64
refill                 int64
doorstep               int64
favday                object
city                  object
dtype: object

In [17]:
print(df2['firstorder'].dt.year.unique())
print(df2['lastorder'].dt.year.unique())

[2013 2011 2010 2012 2009 2008 2014 2015 2017 2016 2018 1904]
[2013 2014 2011 2010 2012 2009 2008 2017 2018 2015 2016 1904]


#### The range of years is supposed to be 2008 to 2018. But we have dates which are of year 1904. So we'll drop the rows which has dates in year 1904

In [18]:
drop_index = df2[(df2['firstorder'].dt.year==1904) | (df2['lastorder'].dt.year==1904)].index
df2.drop(drop_index,axis=0,inplace=True)
df2.reset_index(drop=True,inplace=True)
print(df2.shape)

(30747, 15)


In [19]:
# duplicate values in custid
df2['custid'].value_counts()[df2['custid'].value_counts()>1]

ZA6HL8    2
VHRL89    2
N2JL9X    2
3TXFZX    2
EQTCV3    2
XW9FJZ    2
Name: custid, dtype: int64

#### There are some duplicated values in the column custid. 6 of these customerids are repeated twice in the data. We can drop all these 12 rows due to inconsistent values

In [20]:
df2[df2['custid'].duplicated(keep=False)==True]

custid  retained    created firstorder  lastorder  esent   eopenrate  \
5432   N2JL9X         0 2011-09-21 2011-08-12 2014-02-01      5    0.000000   
7302   N2JL9X         0 2011-09-21 2011-08-12 2014-02-01      1  100.000000   
13726  ZA6HL8         1 2013-09-26 2013-09-26 2013-09-26     32   18.750000   
15600  ZA6HL8         0 2013-09-26 2013-09-26 2013-09-26     13   46.153846   
17179  VHRL89         1 2013-08-10 2013-10-31 2013-10-31     39   17.948718   
17235  XW9FJZ         1 2013-09-10 2013-09-10 2013-11-11     14   14.285714   
20592  VHRL89         0 2013-08-10 2013-10-31 2013-10-31     18    0.000000   
21702  XW9FJZ         1 2013-09-10 2013-09-10 2013-11-11     18   11.111111   
22517  3TXFZX         1 2013-12-12 2013-12-12 2013-12-12     16   25.000000   
22741  EQTCV3         1 2013-10-22 2013-12-14 2013-12-14     25   12.000000   
23266  EQTCV3         1 2013-10-22 2013-12-14 2013-12-14     14   28.571429   
30222  3TXFZX         1 2017-12-13 2017-12-13 2017-12-13     22   50.000000   

       eclickrate  avgorder   ordfreq  paperless  refill  doorstep    favday  \
5432     0.000000     19.45  0.043651          1       0         0    Friday   
7302   100.000000     19.45  0.043651          1       0         0    Friday   
13726    3.125000     40.02  0.000000          1       0         0    Monday   
15600    0.000000     40.02  0.000000          1       1         0    Monday   
17179    7.692308     41.49  0.000000          1       1         1  Saturday   
17235    0.000000     80.72  0.151515          1       0         0   Tuesday   
20592    0.000000     41.49  0.000000          1       0         0   Tuesday   
21702    0.000000    108.84  0.129032          1       0         0   Tuesday   
22517    0.000000     29.00  0.000000          1       0         0  Saturday   
22741    8.000000     40.02  0.000000          1       1         0    Friday   
23266    0.000000     40.02  0.000000          1       1         0    Friday   
30222   13.636364     29.00  0.000000          1       0         0  Saturday   

      city  
5432   BOM  
7302   BOM  
13726  BOM  
15600  BOM  
17179  MAA  
17235  BLR  
20592  MAA  
21702  BLR  
22517  BLR  
22741  BOM  
23266  BOM  
30222  BLR

In [21]:
# Dropping Duplicate values in custid
dup_index = df2[df2['custid'].duplicated(keep=False)==True].index
df3 = df2.drop(dup_index)

# reset the index values
df3.reset_index(drop=True, inplace = True)

# shape of dataset
df3.shape

(30735, 15)

# Data Vectorization

In [22]:
# checking number of unique values in each columns
for col in df3.columns:
    print(col,':',df3[col].nunique())

custid : 30735
retained : 2
created : 2821
firstorder : 2669
lastorder : 2412
esent : 90
eopenrate : 977
eclickrate : 503
avgorder : 9927
ordfreq : 3836
paperless : 2
refill : 2
doorstep : 2
favday : 7
city : 4


##### From above analysis on number of unique values, we can say that
 - custid column has 30741 unique values which is equal to that of no.of rows in our dataset. So we can drop this column, since it doesnot add any information to our dataset
 - The columns which has only less unique values can be considered as categorical columns i.e. columns 'retained','paperless','refill','doorstep','favday' and 'city'
 - Columns 'created, 'firstorder', 'lastorder' are of datetime format
 - Columns 'esent', 'eopenrate', 'eclickrate', 'avgorder', 'ordfreq' are numerical columns

#### Dropping custid column since we have all unique values

In [23]:
df3.drop(['custid'],axis=1,inplace=True)

In [24]:
df3.dtypes

retained               int64
created       datetime64[ns]
firstorder    datetime64[ns]
lastorder     datetime64[ns]
esent                  int64
eopenrate            float64
eclickrate           float64
avgorder             float64
ordfreq              float64
paperless              int64
refill                 int64
doorstep               int64
favday                object
city                  object
dtype: object

In [25]:
# Numerical Columns
num_cols = ['esent', 'eopenrate', 'eclickrate', 'avgorder', 'ordfreq']

In [26]:
# Categorical Columns
cat_cols = ['retained','paperless','refill','doorstep','favday','city']

In [27]:
df4 = df3.copy()
df4.shape

(30735, 14)

# Discrepancies and Modifications

#### There are some discrepancies in the date columns. Practically, Lastorder date should be greater than firstorder date and firstorder date should be greater than created date. So the rows, which doesnot follow the above mentioned rule, are dropped

In [28]:
drop_rows = df4[(df4['firstorder']>df4['lastorder'])|(df4['created']>df4['firstorder'])]
len(drop_rows)

5370

#### The above mentioned discrepancy is found in 5370 rows. So these rows are dropped

In [29]:
df4.drop(drop_rows.index,inplace=True)
df4.reset_index(drop=True,inplace=True)
df4.shape

(25365, 14)

In [30]:
df4.shape

(25365, 14)

# Additional data

#### We are adding 3 more columns that supplements the dataset. The columns are derived from dataset.
- *eopen* - No.of emails opened by the customer
- *tenure* - The tenure of the customer calculated using difference between lastorder date and created date of the customer
- *recency* - No. of days since the last order was placed by the customer (Difference between max_lastorder date and customer's lastorder date)


In [31]:
# eopen 
df4['eopen'] = (df4['esent'] * df4['eopenrate'])/100
df4['eopen'] = df4['eopen'].apply(lambda x: math.ceil(x))

# tenure
df4['tenure'] = df4['lastorder'] - df4['created']
df4['tenure'] = df4['tenure'].apply(lambda x: str(x).split(' ')[0])
df4['tenure'] = df4['tenure'].astype('int64')
df4['tenure'] = df4['tenure']/365

# recency
max_date = df4['lastorder'].max()
df4['recency'] = df4['lastorder'].apply(lambda x: max_date-x)
df4['recency'] = df4['recency'].apply(lambda x: str(x).split(' ')[0])
df4['recency'] = df4['recency'].astype('int64')
df4['recency'] = df4['recency']/365

In [32]:
df4.columns

Index(['retained', 'created', 'firstorder', 'lastorder', 'esent', 'eopenrate',
       'eclickrate', 'avgorder', 'ordfreq', 'paperless', 'refill', 'doorstep',
       'favday', 'city', 'eopen', 'tenure', 'recency'],
      dtype='object')

In [33]:
df4.head()

retained    created firstorder  lastorder  esent   eopenrate  eclickrate  \
0         0 2012-09-28 2013-11-08 2013-11-08     29  100.000000    3.448276   
1         1 2010-12-19 2011-01-04 2014-01-19     95   92.631579   10.526316   
2         0 2010-10-22 2011-03-28 2011-03-28      0    0.000000    0.000000   
3         1 2010-11-27 2010-11-29 2013-01-28     30   90.000000   13.333333   
4         1 2008-11-17 2010-12-10 2014-01-14     46   80.434783   15.217391   

   avgorder   ordfreq  paperless  refill  doorstep     favday city  eopen  \
0     14.52  0.000000          0       0         0     Monday  DEL     29   
1     83.69  0.181641          1       1         1     Friday  DEL     89   
2     54.96  0.000000          0       0         0   Thursday  BOM      0   
3    111.91  0.008850          0       0         0     Monday  BOM     27   
4    175.10  0.141176          1       1         0  Wednesday  DEL     38   

     tenure   recency  
0  1.112329  5.065753  
1  3.087671  4.868493  
2  0.430137  7.684932  
3  2.172603  5.843836  
4  5.161644  4.882192

In [34]:
df4.dtypes

retained               int64
created       datetime64[ns]
firstorder    datetime64[ns]
lastorder     datetime64[ns]
esent                  int64
eopenrate            float64
eclickrate           float64
avgorder             float64
ordfreq              float64
paperless              int64
refill                 int64
doorstep               int64
favday                object
city                  object
eopen                  int64
tenure               float64
recency              float64
dtype: object

In [35]:
# Exporting it to a new file
df4.to_csv('clean_data.csv',index = None,header=True)